### PR TITLE
[M2] Proper port management and refactoring.

### LINKED ISSUE #36: [M2] Identify 'dmsg.Transport' via remote and local pk:port pair.
**Feature description**

Currently `dmsg.Tranport`s are identified by an `uint16` value commonly referred to as the *Transport ID*. On top of this, we also have ports (also represented by an `uint16` value) that is also used for identifying `dmsg.Transport`s.

**Additional context**

Originally, Brandon wanted `dmsg` to have no ports, and we just basically create connections between`dmsg.Client`s. This required us to differentiate the connections from each other, so we know which message is going to where. Hence, Brandon suggested we use a Transport ID (represented with `uint16`).

However, now Brandon wanted to introduce ports. If we had known we wanted ports earlier, we could just have identified the connections via the combination of the local/remote pubKey/port pair (like a TCP connection).

But we didn't know this, so now `dmsg` is also a mess, with both the ideas of Transport IDs and ports.

### PR DESCRIPTION
**Pull request by evanlinjin**
closes #36 

* Introduced netutil.Porter that assigns ports (including ephemeral ports).
* Integrate netutil.Porter to the rest of the codebase.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -304,14 +301,7 @@ func (c *Client) Close() error {
...
-		c.pm.mu.Lock()
-		defer c.pm.mu.Unlock()
-
-		for _, lis := range c.pm.listeners {
-			lis.close()
-		}
 	})
 
-	return nil
+	return c.pm.Close()
```
Comment (Darkren) [inline client.go]: Shouldn't this one be called inside `once.Do`?

File: netutil/porter.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+const (
+	// PorterMinEphemeral is the default minimum ephemeral port.
+	PorterMinEphemeral = uint16(49152)
+)
+
+// Porter reserves ports.
+type Porter struct {
+	sync.RWMutex
+	eph    uint16 // current ephemeral value
```
Comment (Darkren) [inline netutil/porter.go:16]: Why don't we use the `routing.Port` type everywhere we can?

**Review by Darkren — COMMENTED** [review]
I'm not sure, and just have some questions. Also I haven't made i through all code yet

File: netutil/porter.go (hunk 2 above)
Comment (evanlinjin) [reply netutil/porter.go:16]: @Darkren I think that will be ideal too, but we cannot have cyclic dependencies. We will need to move the `Port` type in a sub-module of `dmsg`. That is work for later.

**Review by nkryuchkov — APPROVED** [review]
Great work

**Pull request merged**